# Case_Study_0

# Objective: EDA of Telco_Bill_Data

# Data:
# Steps
* Load the data
* Check # of records and variables
* Check for any Duplicates at row level
* Understand the variables and Write down the generic description of each and every variable
* Produce the summary statistics of both Numerical and categorical variable
* Missing value
* Outliers
* master datasets

# Step1: Load the data

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [46]:
#Read input file
telco_bill = pd.read_excel("C:\\Users\\guptshik\\Desktop\\DVS\\Assignments\\Case_Study_0\\input\\Telco_Bill_data_v1.xlsx")

# Step2: Check columns, num of records and variables

In [26]:
#Get idea about the file by looking at first 5 records
telco_bill.head().T

0  \
CIRCLE                                Karnataka   
HIERARCHY_ID                          206190410   
PARENT_ID                                   NaN   
ACCOUNT_NO                            206190410   
SUBSCR_NO                              14218954   
DEL_NO                               9243148228   
SERVICE_ACTIVE_DT           2006-06-22 00:00:00   
INVOICE_NO                           1225097295   
BILL_SEQUENCE_NUM                            71   
MONTH_YY                                 APR_12   
BILL_PERIOD                                 M25   
BILL_COMPANY             HCL EAI SERVICES LTD.,   
ACCOUNT_CATEGORY                  Top corporate   
MKT_SEGMENT                           TC - COCP   
PRODUCT                             CDMA Mobile   
DATA_USAGE                                    0   
DATA_USAGE_UNITS                              0   
NRC_AMOUNT                                    0   
RC_AMOUNT                                   750   
TAX_AMOUNT                                 61.8   
NEW_CHRGS                                 561.8   
CURRENT_CHRGS                             561.8   
HANDSET_AMOUNT                                0   
UNITCREDIT_AMOUNT                             0   
DISCOUNT                                   -250   
ADJ_AMOUNT                                    0   
ACT_ACTIVE_DT               2006-06-22 00:00:00   
PACKAGE_ID                                54518   
PRICE_PLAN           CMo - Millennium Edition 1   
BILL_ZIP                                 560095   
CUSTOMER_SEGMENT                              S   
STATUS                                   ACTIVE   
OTHERS                                        0   
OTHERS_UNITS                                  0   
REV_RCV_COST_CENTER                   BANGALORE   

                                                               1  \
CIRCLE                                                 Karnataka   
HIERARCHY_ID                                           206197211   
PARENT_ID                                                    NaN   
ACCOUNT_NO                                             206197211   
SUBSCR_NO                                               14232708   
DEL_NO                                                9243264060   
SERVICE_ACTIVE_DT                            2006-06-24 00:00:00   
INVOICE_NO                                            1207231515   
BILL_SEQUENCE_NUM                                             70   
MONTH_YY                                                  APR_12   
BILL_PERIOD                                                  M01   
BILL_COMPANY                                                   .   
ACCOUNT_CATEGORY                               RESIDENTIAL - EBS   
MKT_SEGMENT                           TC - Non COCP Non Top Mgmt   
PRODUCT                                              CDMA Mobile   
DATA_USAGE                                                     0   
DATA_USAGE_UNITS                                               0   
NRC_AMOUNT                                                     0   
RC_AMOUNT                                                    149   
TAX_AMOUNT                                                 24.68   
NEW_CHRGS                                                 224.35   
CURRENT_CHRGS                                             224.35   
HANDSET_AMOUNT                                                 0   
UNITCREDIT_AMOUNT                                              0   
DISCOUNT                                                     -80   
ADJ_AMOUNT                                                     0   
ACT_ACTIVE_DT                                2006-06-23 00:00:00   
PACKAGE_ID                                                617691   
PRICE_PLAN           0610/BOTH/CM/TRF/NU/PO/CMO PPS 99-0610 Plan   
BILL_ZIP                                                  580029   
CUSTOMER_SEGMENT                                               S   
STATUS                           

In [7]:
#Get idea about the file by looking at last 5 records
telco_bill.tail()

CIRCLE  HIERARCHY_ID  PARENT_ID  ACCOUNT_NO  SUBSCR_NO      DEL_NO  \
31155  Karnataka     911845527        NaN   911845527   39342213  7411098935   
31156  Karnataka     911818079        NaN   911818079   39258538  8123023475   
31157  Karnataka     911814457        NaN   911814457   39239175  9035025032   
31158  Karnataka     911798576        NaN   911798576   39221104  8050068409   
31159  Karnataka     911741638        NaN   911741638   39186066  8123093503   

      SERVICE_ACTIVE_DT  INVOICE_NO  BILL_SEQUENCE_NUM MONTH_YY  ...  \
31155        2012-03-21  1238802032                  3   MAY_12  ...   
31156        2012-03-16  1238831122                  3   MAY_12  ...   
31157        2012-03-15  1238831139                  3   MAY_12  ...   
31158        2012-03-14  1239196144                  3   MAY_12  ...   
31159        2012-03-12  1238467118                  3   MAY_12  ...   

      ADJ_AMOUNT ACT_ACTIVE_DT PACKAGE_ID  \
31155        0.0    2012-03-21   629462.0   
31156        0.0    2012-03-16   629462.0   
31157        0.0    2012-03-15   629462.0   
31158        0.0    2012-03-14   629462.0   
31159        0.0    2012-03-12   629462.0   

                                         PRICE_PLAN BILL_ZIP  \
31155  0112/BOTH/GS/TRF/NU/PO/GSM New Speed 99 plan   560096   
31156  0112/BOTH/GS/TRF/NU/PO/GSM New Speed 99 plan   570019   
31157  0112/BOTH/GS/TRF/NU/PO/GSM New Speed 99 plan   560003   
31158  0112/BOTH/GS/TRF/NU/PO/GSM New Speed 99 plan   560102   
31159  0112/BOTH/GS/TRF/NU/PO/GSM New Speed 99 plan   570019   

       CUSTOMER_SEGMENT  STATUS  OTHERS  OTHERS_UNITS  REV_RCV_COST_CENTER  
31155                 N  ACTIVE    0.00             0            BANGALORE  
31156                 Y  ACTIVE  216.05           501               MYSORE  
31157                 Y  TSP_IC  308.71          4584            BANGALORE  
31158                 Y  ACTIVE    0.00             0            BANGALORE  
31159                 Y  ACTIVE  191.42           696               MYSORE  

[5 rows x 35 columns]

In [6]:
#Get info about how many row and columns are there in the data (31160 rows and 35 columns)
telco_bill.shape

(31160, 35)

In [8]:
#Get list of columns
telco_bill.columns

Index(['CIRCLE', 'HIERARCHY_ID', 'PARENT_ID', 'ACCOUNT_NO', 'SUBSCR_NO',
       'DEL_NO', 'SERVICE_ACTIVE_DT', 'INVOICE_NO', 'BILL_SEQUENCE_NUM',
       'MONTH_YY', 'BILL_PERIOD', 'BILL_COMPANY', 'ACCOUNT_CATEGORY',
       'MKT_SEGMENT', 'PRODUCT', 'DATA_USAGE', 'DATA_USAGE_UNITS',
       'NRC_AMOUNT', 'RC_AMOUNT', 'TAX_AMOUNT', 'NEW_CHRGS', 'CURRENT_CHRGS',
       'HANDSET_AMOUNT', 'UNITCREDIT_AMOUNT', 'DISCOUNT', 'ADJ_AMOUNT',
       'ACT_ACTIVE_DT', 'PACKAGE_ID', 'PRICE_PLAN', 'BILL_ZIP',
       'CUSTOMER_SEGMENT', 'STATUS', 'OTHERS', 'OTHERS_UNITS',
       'REV_RCV_COST_CENTER'],
      dtype='object')

In [9]:
# check for detailed info about data types, Null or non-null columns
telco_bill.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31160 entries, 0 to 31159
Data columns (total 35 columns):
CIRCLE                 31160 non-null object
HIERARCHY_ID           31160 non-null int64
PARENT_ID              7998 non-null float64
ACCOUNT_NO             31160 non-null int64
SUBSCR_NO              31160 non-null int64
DEL_NO                 31160 non-null int64
SERVICE_ACTIVE_DT      31160 non-null datetime64[ns]
INVOICE_NO             31160 non-null int64
BILL_SEQUENCE_NUM      31160 non-null int64
MONTH_YY               31160 non-null object
BILL_PERIOD            31160 non-null object
BILL_COMPANY           31156 non-null object
ACCOUNT_CATEGORY       31160 non-null object
MKT_SEGMENT            31160 non-null object
PRODUCT                31160 non-null object
DATA_USAGE             31160 non-null float64
DATA_USAGE_UNITS       31160 non-null int64
NRC_AMOUNT             31160 non-null int64
RC_AMOUNT              31160 non-null float64
TAX_AMOUNT             31160 non-n

# Step3: Check for duplicates and remove them

In [13]:
# Check if we have duplicates. Results shows we have duplicates
df_dup = telco_bill.duplicated()
print(sum(df_dup==True))
print(sum(df_dup==False))

21
31139


In [47]:
#remove duplicates
telcoBill = telco_bill.drop_duplicates()

In [15]:
# Check if we have duplicates. Results shows after removing duplicates, new DF does not have duplicates
df_dup = telcoBill.duplicated()
print(sum(df_dup==True))
print(sum(df_dup==False))

0
31139


In [48]:
telcoBill.shape

(31139, 35)

# Step4: Understand the variables and Write down the generic description of each and every variable

In [100]:
telcoBill.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31139 entries, 0 to 31159
Data columns (total 38 columns):
CIRCLE                 31139 non-null object
HIERARCHY_ID           31139 non-null int64
PARENT_ID              7994 non-null float64
ACCOUNT_NO             31139 non-null int64
SUBSCR_NO              31139 non-null int64
DEL_NO                 31139 non-null int64
SERVICE_ACTIVE_DT      31139 non-null datetime64[ns]
INVOICE_NO             31139 non-null int64
BILL_SEQUENCE_NUM      31139 non-null int64
MONTH_YY               31139 non-null object
BILL_PERIOD            31139 non-null object
BILL_COMPANY           31135 non-null object
ACCOUNT_CATEGORY       31139 non-null object
MKT_SEGMENT            31139 non-null object
PRODUCT                31139 non-null object
DATA_USAGE             31139 non-null float64
DATA_USAGE_UNITS       31139 non-null int64
NRC_AMOUNT             31139 non-null int64
RC_AMOUNT              31139 non-null float64
TAX_AMOUNT             31139 non-n

# Data Dictionary or description or Attributes
# Numerical variables:
1.	HIERARCHY_ID-->
2.	PARENT_ID-->
3.	ACCOUNT_NO-->
4.	SUBSCR_NO-->
5.	DEL_NO-->
6.	INVOICE_NO-->
7.	BILL_SEQUENCE_NUM-->
8.	DATA_USAGE-->
9.	DATA_USAGE_UNITS-->
10.	NRC_AMOUNT-->
11.	RC_AMOUNT-->
12.	TAX_AMOUNT-->
13.	NEW_CHRGS-->
14.	CURRENT_CHRGS-->
15.	HANDSET_AMOUNT-->
16.	UNITCREDIT_AMOUNT-->
17.	DISCOUNT-->
18.	ADJ_AMOUNT-->
19.	PACKAGE_ID-->
20.	BILL_ZIP-->
21.	OTHERS-->
22.	OTHERS_UNITS-->


# Categorical Variables: (text data, Binary data yes/no, True/False, 0/1 etc)
1.	CIRCLE-->
2.	SERVICE_ACTIVE_DT-->
3.	MONTH_YY-->
4.	BILL_PERIOD-->
5.	BILL_COMPANY-->
6.	ACCOUNT_CATEGORY-->
7.	MKT_SEGMENT-->
8.	PRODUCT-->
9.	ACT_ACTIVE_DT-->
10.	PRICE_PLAN-->
11.	CUSTOMER_SEGMENT-->
12.	STATUS-->
13.	REV_RCV_COST_CENTER-->


# Step5: Summary Statistics 
1. For numerical data--> df["col_name"].describe()
2. For categorical data--> df["col_name"].value_counts()/len(df)*100 

In [101]:
telcoBill.describe().T

count          mean           std           min  \
HIERARCHY_ID       31139.0  8.014430e+08  2.482739e+08  2.000089e+08   
PARENT_ID           7994.0  6.847059e+08  3.134651e+08  2.052738e+08   
ACCOUNT_NO         31139.0  7.667885e+08  2.816160e+08  2.000089e+08   
SUBSCR_NO          31139.0  3.166616e+07  6.794137e+06  1.857437e+06   
DEL_NO             31139.0  8.215899e+09  2.280726e+09  1.008000e+09   
INVOICE_NO         31139.0  1.226770e+09  1.086179e+07  1.207171e+09   
BILL_SEQUENCE_NUM  31139.0  2.137580e+01  1.941593e+01  1.000000e+00   
DATA_USAGE         31139.0  2.165376e+02  9.434007e+02  0.000000e+00   
DATA_USAGE_UNITS   31139.0  4.231577e+02  1.764224e+03  0.000000e+00   
NRC_AMOUNT         31139.0  4.216931e+00  2.908160e+02  0.000000e+00   
RC_AMOUNT          31139.0  3.066132e+03  5.750978e+04  0.000000e+00   
TAX_AMOUNT         31139.0  2.126662e+02  2.540486e+03 -1.879985e+05   
NEW_CHRGS          31139.0  2.328050e+03  2.806139e+04  0.000000e+00   
CURRENT_CHRGS      31139.0  1.908206e+03  2.406475e+04 -2.058341e+06   
HANDSET_AMOUNT     31139.0  1.862616e-01  5.343009e+00  0.000000e+00   
UNITCREDIT_AMOUNT  31139.0 -7.528659e+00  1.756523e+02 -2.124443e+04   
DISCOUNT           31139.0 -1.594394e+03  3.647938e+04 -5.471795e+06   
ADJ_AMOUNT         31139.0 -3.768136e+02  2.692090e+04 -4.191826e+06   
PACKAGE_ID         30989.0  5.840618e+05  1.390624e+05  1.090000e+02   
BILL_ZIP           31139.0  5.608897e+05  7.338744e+03  5.600000e+02   
OTHERS             31139.0  4.402247e+02  5.416676e+03  0.000000e+00   
OTHERS_UNITS       31139.0  4.376957e+03  9.254116e+04  0.000000e+00   

                            25%           50%           75%           max  
HIERARCHY_ID       9.049237e+08  9.081479e+08  9.099571e+08  9.804977e+08  
PARENT_ID          2.085882e+08  9.044682e+08  9.049237e+08  9.804977e+08  
ACCOUNT_NO         9.048178e+08  9.080802e+08  9.098945e+08  9.146603e+08  
SUBSCR_NO          2.791352e+07  3.314639e+07  3.744356e+07  4.130682e+07  
DEL_NO             8.066969e+09  9.243209e+09  9.243393e+09  1.860266e+10  
INVOICE_NO         1.215873e+09  1.226890e+09  1.235972e+09  1.291591e+09  
BILL_SEQUENCE_NUM  6.000000e+00  1.600000e+01  3.100000e+01  1.150000e+02  
DATA_USAGE         0.000000e+00  0.000000e+00  5.000000e+00  5.543300e+04  
DATA_USAGE_UNITS   0.000000e+00  0.000000e+00  1.000000e+01  1.108660e+05  
NRC_AMOUNT         0.000000e+00  0.000000e+00  0.000000e+00  4.000000e+04  
RC_AMOUNT          1.245700e+02  4.990000e+02  7.500000e+02  7.793279e+06  
TAX_AMOUNT         1.561000e+01  5.812000e+01  9.270000e+01  2.748762e+05  
NEW_CHRGS          1.449400e+02  5.337100e+02  8.427000e+02  2.608419e+06  
CURRENT_CHRGS      1.430300e+02  5.337100e+02  8.427000e+02  2.498794e+06  
HANDSET_AMOUNT     0.000000e+00  0.000000e+00  0.000000e+00  2.000000e+02  
UNITCREDIT_AMOUNT  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
DISCOUNT          -5.893950e+02 -1.655000e+02 -3.335000e+01  0.000000e+00  
ADJ_AMOUNT         0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
PACKAGE_ID         6.123360e+05  6.181080e+05  6.221220e+05  6.308570e+05  
BILL_ZIP           5.600290e+05  5.600480e+05  5.600760e+05  6.000010e+05  
OTHERS             0.000000e+00  0.000000e+00  2.985000e+02  5.189340e+05  
OTHERS_UNITS       0.000000e+00  0.000000e+00  7.650000e+02  7.413343e+06

# Initial insight of numerical data from summary statistics of the data df.describe().T
* After removing duplicates it has 35 cols and 31139 records
* It has 5 columns with missing values
 * PARENT_ID has 23145 NaN
 * BILL_COMPANY has 4 NaN
 * PACKAGE_ID has 150 NaN
 * PRICE_PLAN has 150 NaN
 * CUSTOMER_SEGMENT has 280 NaN

# Binning or Bucketing continuous data

In [102]:
telcoBill['SERVICE_ACTIVE_DT1'] = pd.cut(telcoBill['SERVICE_ACTIVE_DT'], 8)
telcoBill['ACT_ACTIVE_DT1'] = pd.cut(telcoBill['ACT_ACTIVE_DT'], 8)

C:\Users\guptshik\AppData\Local\Continuum\anaconda2\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\guptshik\AppData\Local\Continuum\anaconda2\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [74]:
def summary_stats_categorical_variable(df_col,df):
    print ('***************************************')
#     print("*****Summary_Statistics of col ****",df_col,"****" )
    print (df[i].value_counts()/len(df)*100)
    print ('***************************************')
######################################################################################

sel_cols = ["CIRCLE", "SERVICE_ACTIVE_DT1", "MONTH_YY", "BILL_PERIOD", "BILL_COMPANY", "ACCOUNT_CATEGORY", "MKT_SEGMENT", "PRODUCT", "ACT_ACTIVE_DT1", "PRICE_PLAN", "CUSTOMER_SEGMENT", "STATUS", "REV_RCV_COST_CENTER"]
print ('***************************************')
for i in sel_cols:
    print (i)
    summary_stats_categorical_variable(i,telcoBill)

***************************************
CIRCLE
***************************************
Karnataka         99.951829
NPLC-Karnataka     0.041748
Tamil Nadu         0.006423
Name: CIRCLE, dtype: float64
***************************************
SERVICE_ACTIVE_DT1
***************************************
(2011-05-04 18:00:00, 2012-07-05]                    46.870484
(2010-03-03 12:00:00, 2011-05-04 18:00:00]           21.802242
(2008-12-31 06:00:00, 2010-03-03 12:00:00]           18.449533
(2007-10-31, 2008-12-31 06:00:00]                     6.615498
(2006-08-29 18:00:00, 2007-10-31]                     3.233887
(2005-06-28 12:00:00, 2006-08-29 18:00:00]            2.533800
(2004-04-27 06:00:00, 2005-06-28 12:00:00]            0.366100
(2003-02-21 13:58:04.800000, 2004-04-27 06:00:00]     0.128456
Name: SERVICE_ACTIVE_DT1, dtype: float64
***************************************
MONTH_YY
***************************************
APR_12    54.426924
MAY_12    44.892257
MAR_12     0.578053
JUL_12 

In [103]:
plan = telcoBill.PRICE_PLAN.str.split(pat='/',expand=True).rename(columns = {0: "P0",1:"P1", 2: "P2", 3: "P3", 4: "P4", 5: "P5", 6: "P6", 7: "P7"})
sel_cols = ["P0", "P1", "P2", "P3", "P4", "P5", "P6", "P7" ]
print ('***************************************')
for i in sel_cols:
    print (i)
    summary_stats_categorical_variable(i,plan)

***************************************
P0
***************************************
0710                                      13.214939
0311                                      10.870612
1208                                       7.415139
1209                                       5.478660
0709                                       5.016218
0112                                       4.987315
0509                                       3.429783
1109                                       3.127910
1210                                       2.906323
0110                                       2.883843
0609                                       2.543434
0106                                       2.517743
0811                                       2.456726
0611                                       2.267253
1010                                       2.228716
1110                                       2.064935
0608                                       1.644240
0310                             

# Step6: Check for missing values
* df.isnull().sum()/len(df)*100

In [124]:
def identify_missing(df):
    df1 = df.isnull().sum()/len(df)*100
    df = df1[df1 > 0]
    print("******************Columns with Missing values and % of missing values ******************")
    print (df)
    
identify_missing(telcoBill)

******************Columns with Missing values and % of missing values ******************
PARENT_ID           74.328013
BILL_COMPANY         0.012846
PACKAGE_ID           0.481711
PRICE_PLAN           0.481711
CUSTOMER_SEGMENT     0.899194
PRICE_PLAN1          0.481711
dtype: float64


# Step7: Identify outliers
* In general outliers can be detected by using the 4 meithods
    * boxlpot
    * qqplot
    * IQR (Most widely used)
    * z-score

1. Outliers are of 2 types
    * LO
    * UO
* L.O. (lower outliers)= Q1 - 3 * IQR (this is practically)
* U.O. (upper outliers)= Q3 + 3 * IQR (this is practically)
* iqr = Q3-Q1
* Mild outliers: for research mild outliers (we wont consider in real time). If huge diff b/w 50% to 75%, 25% to 50%
* Wild outliers(LO, UO) mu - 3 sigma to mu + 3 sigma 
* B/c of Mild outliers we have data NOT normally distributed hence we apply variable transformation technique
    * square
    * sqrt
    * Log
    * Inverse


In [125]:
import flask